libbeat/processors/dns/cache.go

```diff
@@ -0,0 +1,194 @@
+}
+
+func (c *ptrCache) evict() {
+	var key string
+	for k := range c.data {
+		key = k
+		break
+	}
+	delete(c.data, key)
+}
+
+func (c *ptrCache) get(now time.Time, key string) *PTR {
+	r, found := c.data[key]
+	if found && !r.IsExpired(now) {
+		return &PTR{r.host, uint32(r.expires.Sub(now) / time.Second)}
+	}
+	return nil
+}
+
+type failureRecord struct {
+	error
+	expires time.Time
+}
+
+func (r failureRecord) IsExpired(now time.Time) bool {
+	return now.After(r.expires)
+}
+
+type failureCache struct {
+	data       map[string]failureRecord
+	maxSize    int
+	failureTTL time.Duration
+	stats      cacheStats
+}
+
+func (c *failureCache) set(now time.Time, key string, err error) {
+	if len(c.data) >= c.maxSize {
+		c.evict()
+	}
+
+	c.data[key] = failureRecord{
+		error:   err,
+		expires: now.Add(c.failureTTL),
+	}
+}
+
+func (c *failureCache) evict() {
+	var key string
+	for k := range c.data {
+		key = k
+		break
+	}
+	delete(c.data, key)
+}
+
+func (c *failureCache) get(now time.Time, key string) error {
+	r, found := c.data[key]
+	if found && !r.IsExpired(now) {
+		return r.error
+	}
+	return nil
+}
+
+// PTRLookupCache is a cache for storing and retrieving the results of
+// reverse DNS queries. It caches the results of queries regardless of their
+// outcome (success or failure).
+type PTRLookupCache struct {
+	success    *ptrCache
+	failure    *failureCache
+	failureTTL time.Duration
+	resolver   PTRResolver
+	log        Logger
+	stats      cacheStats
+}
+
+type cacheStats struct {
+	Hit  *monitoring.Int
+	Miss *monitoring.Int
+}
+
+// Logger logs debug messages.
+type Logger interface {
+	Debugw(msg string, keysAndValues ...interface{})
+}
+
+// NewPTRLookupCache returns a new cache.
+func NewPTRLookupCache(reg *monitoring.Registry, l Logger, conf CacheConfig, resolver PTRResolver) *PTRLookupCache {
+	c := &PTRLookupCache{
+		success: &ptrCache{
+			data:    make(map[string]ptrRecord, conf.SuccessCache.InitialCapacity),
+			maxSize: max(100, max(conf.SuccessCache.InitialCapacity, conf.SuccessCache.MaxCapacity)),
+		},
+		failure: &failureCache{
+			data:       make(map[string]failureRecord, conf.FailureCache.InitialCapacity),
+			maxSize:    max(100, max(conf.FailureCache.InitialCapacity, conf.FailureCache.MaxCapacity)),
+			failureTTL: conf.FailureCache.TTL,
+		},
+		resolver: resolver,
+		log:      l,
+		stats: cacheStats{
+			Hit:  monitoring.NewInt(reg, "hits"),
+			Miss: monitoring.NewInt(reg, "misses"),
+		},
+	}
+
+	return c
+}
+
+// LookupPTR performs a reverse lookup on the given IP address. A cached result
+// will be returned if it is contained in the cache, otherwise a lookup is
+// performed.
+func (c PTRLookupCache) LookupPTR(ip string) (*PTR, error) {
+	now := time.Now()
+
+	ptr := c.success.get(now, ip)
+	if ptr != nil {
+		c.stats.Hit.Inc()
+		return ptr, nil
+	}
+
+	err := c.failure.get(now, ip)
+	if err != nil {
+		c.stats.Hit.Inc()
+		return nil, err
+	}
+	c.stats.Miss.Inc()
+
+	ptr, err = c.resolver.LookupPTR(ip)
```

Comment:
Might be nice to have a stat for actual DNS queries sent to help users determine if their cache size is appropriate. They could check if increasing the cache size diminishes the number of queries sent.

Reply:
Yes, cache misses is an even better statistic.

I think I missed that we report those stats already, but it does make sense that cache misses would be a good stat.